lib/array.js

```diff
@@ -38,6 +38,60 @@ const map = (
   }
 };
 
+const NS_PER_SEC = 1e7;
+
+const asyncMap = (
+  // Non blocking map
+  items, // array, incoming
+  fn, // function, (value, index) => {}
+  {
+    min = 5,
+    max = 30,
+  } = {}
+) => {
+  let count = 0;
+  let setTimeoutCount = 0;
+
+  let maxTime = 0;
+  let minTime = Infinity;
+  let sumTime = 0;
+
+  let begin = process.hrtime();
+
+  const halfRange = (max - min) / 2;
+  const middle = min + halfRange;
+
+  let iterMax = middle;
+
+  const countNecessaryNumber = () => {
+    let time = process.hrtime(begin);
+    time = time[0] + NS_PER_SEC + time[1];
+
+    maxTime = Math.max(time, maxTime);
+    minTime = Math.min(time, minTime);
+    sumTime += time;
+
+    const avarageTime = sumTime / ++setTimeoutCount;
+    const timeRange = time <= avarageTime ?
+      avarageTime - minTime : maxTime - avarageTime;
+    const percent = (time - avarageTime) / timeRange;
+
+    return middle + halfRange * percent;
+  };
+
+  const next = () => {
+    while (count < iterMax) fn(items[count], ++count);
```

Comment:
You are using prefix increment here, meaning that you will increment the `count` before passing it as the second argument to the `fn()`, but not before accessing the `items` array (for example, `fn(items[0], 1)`).